sympy/external/gmpy.py

```diff
@@ -179,6 +191,7 @@
     legendre = python_legendre
     jacobi = python_jacobi
     kronecker = python_kronecker
+    iroot = python_iroot
```

Comment:
I went for a bit of a quick fix when adding ground types = flint and just added the python versions of these functions.

Now that we have flint as a possibility it is worth investigating what things it has e.g.:
```python
In [12]: import flint

In [13]: [n for n in dir(flint.fmpz) if not n.startswith('_')]
Out[13]: 
['bell_number',
 'bin_uiui',
 'bit_length',
 'denominator',
 'divisor_sigma',
 'euler_number',
 'euler_phi',
 'fac_ui',
 'factor',
 'factor_smooth',
 'fib_ui',
 'gcd',
 'height_bits',
 'is_perfect_power',
 'is_prime',
 'is_probable_prime',
 'isqrt',
 'jacobi',
 'lcm',
 'moebius_mu',
 'numerator',
 'partitions_p',
 'primorial_ui',
 'repr',
 'rising',
 'root',
 'sqrtmod',
 'sqrtrem',
 'stirling_s1',
 'stirling_s2',
 'str']
```
In this case the `fmpz.root` method is what we want:
```python
In [15]: flint.fmpz(10).root(2)
Out[15]: 3
```
There are some cases that might need to be handled specially. You can see the python-flint code for this here:
https://github.com/flintlib/python-flint/blob/99fa557c8c358cfeec307c6ed11c1315893a9ca9/src/flint/fmpz.pyx#L871-L879
The docs for the relevant Flint function are here:
https://flintlib.org/doc/fmpz.html#c.fmpz_root
Note that in the Flint docs when it says an "exception" what it means is a complete crash i.e. `abort` in C. From Python we should avoid calling any function in such a way that would trigger Flint to raise an "exception". Really it should be python-flint's job to do that but sometimes there are cases that are missed so it is worth checking edge cases.

Reply:
Note that flint has a similar limitation to gmpy2:
```python
In [25]: flint.fmpz(10**100).root(2**63-1)
Out[25]: 1

In [26]: flint.fmpz(10**100).root(2**63)
---------------------------------------------------------------------------
OverflowError 
```
If we suppose that `x = 2**m` then how big does `m` need to be such that `iroot(x, 2**63)` is anything other than 1? So we want to know when:
```python
(2**m)**(1/2**63)) >= 2
(2**(m/2**63)) >= 2
m/2**63 >= 1
m >= 2**63 
```
This would require `x` to be a `2**63`-bit number which you could maybe store in memory if you had about 10 million terabytes of RAM.

On a 32 bit machine it is possible that both flint and gmpy2 would have a lower limit like `2**31` which only needs about 250 megabytes so is plausible for a modern 32 bit machine (although 32 bit is uncommon these days).

Reply:
Set a threshold and call `python_iroot` if it is higher than that

Reply:
Other functions could be written as follows

```python
def sqrt(x):
    return flint.fmpz(x).isqrt()

def is_square(x):
    return flint.fmpz(x).sqrtrem()[1] == 0

def sqrtrem(x):
    return flint.fmpz(x).sqrtrem()

def gcd(*args):
    return reduce(flint.fmpz.gcd, args, flint.fmpz(0))

def lcm(*args):
    return reduce(flint.fmpz.lcm, args, flint.fmpz(1))

def jacobi(x, y):
    if y <= 0 or not y % 2:
        raise ValueError("y should be an odd positive integer")
    return flint.fmpz(x).jacobi(y)
```